From b3457276511c6cb748618a0cfd9dca4fdaaf4679 Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Fri, 24 Apr 2020 11:36:45 -0700
Subject: [PATCH] gold: Handle local IFUNC symbol for IBT PLT

	PR gold/25872
	* x86_64.cc
	(Output_data_plt_x86_64_ibt<size>::do_address_for_local): Handle
	IFUNC symbol.
---
 gold/x86_64.cc | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/gold/x86_64.cc b/gold/x86_64.cc
index 1442a009d7e..284809c719f 100644
--- a/gold/x86_64.cc
+++ b/gold/x86_64.cc
@@ -2288,7 +2288,12 @@ Output_data_plt_x86_64_ibt<size>::do_address_for_local(const Relobj* object,
 						 unsigned int r_sym)
 {
   // Convert the PLT offset into an APLT offset.
-  unsigned int plt_offset = ((object->local_plt_offset(r_sym) - plt_entry_size)
+  const Sized_relobj_file<size, false>* sized_relobj =
+    static_cast<const Sized_relobj_file<size, false>*>(object);
+  const Symbol_value<size>* psymval = sized_relobj->local_symbol(r_sym);
+  unsigned int plt_offset = ((object->local_plt_offset(r_sym)
+			      - (psymval->is_ifunc_symbol()
+				 ? 0 : plt_entry_size))
 			     / (plt_entry_size / aplt_entry_size));
   return (this->address()
 	  + this->aplt_offset_
-- 
2.26.0

